crates/core/src/metadata_store/providers/etcd.rs

```diff
@@ -0,0 +1,286 @@
+    /// precondition is not met, then the operation returns a [`WriteError::PreconditionViolation`].
+    ///
+    /// NOTE: this implementation disregard that version attached on the value and depends only on the
+    /// auto version provided by etcd
+    ///
+    /// it's up to the caller of this function to make sure that both versions are in sync.
+    async fn put(
+        &self,
+        key: ByteString,
+        value: VersionedValue,
+        precondition: Precondition,
+    ) -> Result<(), WriteError> {
+        let mut client = self.client.kv_client();
+        match precondition {
+            Precondition::None => {
+                client.put(key.into_bytes(), value.value, None).await?;
+            }
+            Precondition::DoesNotExist => {
+                if !self
+                    .put_not_exist(&mut client, key.into_bytes(), value.value)
+                    .await?
+                {
+                    // pre condition failed.
+                    return Err(WriteError::FailedPrecondition(
+                        "key-value pair exists".into(),
+                    ));
+                };
+            }
+            Precondition::MatchesVersion(version) => {
+                if !self
+                    .put_with_version(
+                        &mut client,
+                        key.into_bytes(),
+                        value.value,
+                        u32::from(version) as i64,
+                    )
+                    .await?
+                {
+                    return Err(WriteError::FailedPrecondition(
+                        "key version mismatch".into(),
+                    ));
+                };
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Deletes the key-value pair for the given key following the provided precondition. If the
+    /// precondition is not met, then the operation returns a [`WriteError::PreconditionViolation`].
+    async fn delete(&self, key: ByteString, precondition: Precondition) -> Result<(), WriteError> {
+        let mut client = self.client.kv_client();
+        match precondition {
+            Precondition::None => {
+                client.delete(key.into_bytes(), None).await?;
+            }
+            Precondition::DoesNotExist => {
+                if !self
+                    .put_not_exist(&mut client, key.into_bytes(), Vec::default().into())
+                    .await?
+                {
+                    // pre condition failed.
+                    return Err(WriteError::FailedPrecondition(
+                        "key-value pair exists".into(),
+                    ));
+                };
+            }
+            Precondition::MatchesVersion(version) => {
+                if !self
+                    .put_with_version(
+                        &mut client,
+                        key.into_bytes(),
+                        Vec::default().into(),
```

Comment:
Sorry for giving you bad advice. I think it is ok to implement the `delete` operation by deleting the complete entry. If users of the `MetadataStore` want to keep the version around, then they need to leave a tombstone behind.